### PR TITLE
Value replacer fix - Optional

### PR DESCRIPTION
**Pull request by davidboren**
So, I wanted to ask a quick question about value replacers default behavior and propose the modification found in this branch.  Are we okay with the following code producing NAs in prediction?

```R
###TRAIN
train_df <- data.frame(x = factor(c("1", "2", "3", "4", "5", "5")), y = factor(c("a", "b", NA,"c", "d", "d")))
mb <- mungebits:::mungebit(syberiaMungebits::value_replacer)
mp <- mungebits::mungeplane(train_df)
mb$run(mp, is.factor, list(list(NA, "Missing")))
###PREDICT
pred_df <- data.frame(x = factor(c("2", "3", NA, "6")), y = factor(c("d", "b", NA, "d")))
mp <- mungebits::mungeplane(pred_df)
mb$run(mp, is.factor, list(list(NA, "Missing")))
print(mp$data)
```

Basically, the current behavior of value_replacer only creates the new level when there is a need (corresponding level-to-replace in training data).  In prediction it transforms every level that it hasn't seen before into NA.
This is kind of okay if we run group_minor_levels immediately after and allow for group_new_levels = TRUE, where it will transform those NAs into 'Other,' but often people run this mungebit with the options "exclude = c('NA', 'Missing')"  Below is an example where running group_minor_levels immediately afterwards alleviates the problem:

```R
###TRAIN
train_df <- data.frame(x = factor(c("1", "2", "2", "2", "2")), y = factor(c("a", "b", "b", "b", NA)))
mb <- mungebits:::mungebit(syberiaMungebits::value_replacer)
mp <- mungebits::mungeplane(train_df)
mb$run(mp, is.factor, list(list(NA, "Missing")))
mb2 <- syberia_project()$resource('lib/mungebits/group_minor_levels')$value()
mb2$run(mp, is.factor, min_pct = 0.3, group_new_levels = TRUE, exclude = c("Missing"))
###PREDICT
pred_df <- data.frame(x = factor(c("2", "3", NA, "6")), y = factor(c("d", "b", NA, "d")))
mp <- mungebits::mungeplane(pred_df)
mb$run(mp, is.factor, list(list(NA, "Missing")))
print(mp$data)
mb2$run(mp, is.factor, min_pct = 0.3, group_new_levels = TRUE, exclude = c("Missing"))
print(mp$data)
```

Currently the fix in this branch only addresses NA's from levels defined in value_replacer.  If the replacement level is specified it will leave it in regardless of whether it is present in training.  It does not leave new levels alone, as we might want to do, in order to avoid them being excluded by group_minor_levels calls made with "exclude = c('NA')"
Mind you, this is entirely unnecessary if group_minor_levels is performed immediately without exclude option "exclude = c('NA')"  The only reason I'm worried is because this behavior is not an obvious 'feature' of value_replacer.  It does not follow from it's name that value_replacer is a transformation mungebit turning levels not present during training into NA.

### CODE REVIEW

**Comment by davidboren** [comment]
Okay this was bugging me too much so I just added the functionality of leaving new levels alone.  I don't think they should be flushed to NAs in this mungebit, since it's unexpected behavior.  The group_minor_levels should take care of them explicitely with the group_new_levels argument.

**Comment by abarnw** [comment]
@robertzk can you please review and merge this, this is essential to get @peherman's scikit random forest working.

**Comment by peterhurford** [comment]
No test changes?

File: R/value_replacer.r [hunk 1]
```diff
@@ -18,6 +18,13 @@ value_replacer_fn <- function(x, values_map) {
   unnamed_indices <- names(values_map) == ""
   if (is.null(unnamed_indices) || length(unnamed_indices) == 0)
     unnamed_indices <- TRUE
+  if (is_factor) { 
+    if (!exists('inputs') || !'levels' %in% names(inputs)) {
+      rep_levels_unnamed <- unlist(lapply(values_map[unnamed_indices], function(value_map) value_map[[2]]))
```
Comment (robertzk) [inline R/value_replacer.r]: Can you try to do this without `unlist`? Maybe `c(recursive = TRUE, ...)`?

**Comment by davidboren** [comment]
Okay, agreed.  New tests added and unlist is replaced with c(recursive = T, ...).  I can't seem to find any example where unlist results in something different than c(recursive = T,...), even when using a whole bunch of nested lists of different types.  Is it an optimization thing or can it actually give different results?

File: inst/tests/test-value_replacer.r [hunk 2]
```diff
@@ -100,5 +100,39 @@ test_that("it restores factors correctly", {
...
+
+  ###PREDICT
+  pred_df <- data.frame(x = factor(c("2", "a", NA)), y = factor(c("a", "b", NA)))
+  mp <- mungebits::mungeplane(pred_df)
+  mb$run(mp, is.factor, list(list(NA, "Missing"), list("a", "was_a")))
+  
+  expected_df <- data.frame(x = factor(c("2", "was_a", "Missing")), y = factor(c("was_a", "b", "Missing")))
+  expect_equal(as.character(mp$data[[1]]),as.character(expected_df[[1]]))
+  expect_equal(as.character(mp$data[[2]]),as.character(expected_df[[2]]))
+})
```
Comment (peterhurford) [inline inst/tests/test-value_replacer.r]: Nice tests :+1:

**Comment by davidboren** [comment]
Good to merge?

**Pull request merged**